util/downloader/downloader.go

```diff
@@ -182,83 +183,68 @@ func (d *Downloader) downloadFile(ctx context.Context, out *os.File, stats *Stat
 		_ = resp.Body.Close()
 	}()
 
-	buffer := make([]byte, 32*1024)
-	hasher := sha256.New()
-
-	if err := d.updateHasherWithExistingData(stats.Downloaded, hasher); err != nil {
-		return err
-	}
-
-	return d.writeToFile(ctx, resp, out, buffer, hasher, stats)
-}
-
-func (d *Downloader) createRequest(ctx context.Context, downloaded int64) (*http.Request, error) {
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, d.url, http.NoBody)
-	if err != nil {
-		return nil, ErrNewRequest
-	}
-	if downloaded > 0 {
-		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", downloaded))
+	if resp.StatusCode != http.StatusPartialContent && resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("got http response %s from %s: %w", resp.Status, d.url, err)
 	}
 
-	return req, nil
-}
+	buf := make([]byte, 32*1024) // 32KB buffer for reading the response body
+	offset := c.start
+	for {
+		n, err := resp.Body.Read(buf)
+		if n > 0 {
+			d.mu.Lock()
+			for written := 0; written < n; {
+				w, err := out.WriteAt(buf[written:n], offset+int64(written))
+				if err != nil {
+					d.mu.Unlock()
 
-func (d *Downloader) updateHasherWithExistingData(downloaded int64, hasher io.Writer) error {
-	if downloaded > 0 {
-		existingFile, err := os.Open(d.filePath)
-		if err != nil {
-			return ErrOpenFileExists
+					return ErrFileWriting
+				}
+				written += w
+			}
+			offset += int64(n)
+			d.downloaded += int64(n)
+			d.updateStats(d.downloaded, totalSize)
+			d.mu.Unlock()
 		}
-		defer func() {
-			_ = existingFile.Close()
-		}()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
 
-		if _, err := io.CopyN(hasher, existingFile, downloaded); err != nil {
-			return ErrCopyExistsFileData
+			return fmt.Errorf("error reading body from %s: %w", d.url, err)
 		}
 	}
 
 	return nil
 }
 
-func (d *Downloader) writeToFile(ctx context.Context, resp *http.Response, out *os.File, buffer []byte,
-	hasher hash.Hash, stats *Stats,
-) error {
-	for {
-		select {
-		case <-ctx.Done():
-			d.stop()
-
-			return ctx.Err()
-		default:
-			n, err := resp.Body.Read(buffer)
-			if n > 0 {
-				if _, err := out.Write(buffer[:n]); err != nil {
-					return ErrFileWriting
-				}
-
-				if _, err := hasher.Write(buffer[:n]); err != nil {
-					return ErrFileWriting
-				}
+func (d *Downloader) updateStats(downloaded, totalSize int64) {
+	stats := Stats{
+		Downloaded: downloaded,
+		TotalSize:  totalSize,
+		Percent:    float64(downloaded) / float64(totalSize) * 100,
+	}
+	d.statsCh <- stats
+}
 
-				stats.Downloaded += int64(n)
-				stats.Percent = float64(stats.Downloaded) / float64(stats.TotalSize) * 100
-				d.statsCh <- *stats
-			}
-			if err != nil {
-				if err == io.EOF {
-					return d.finalizeDownload(hasher, stats)
-				}
+func (d *Downloader) finalizeDownload(stats *Stats) error {
+	// Recalculate the hash by re-reading the entire file
+	out, err := os.Open(d.filePath)
+	if err != nil {
+		return ErrOpenFileExists
```

Comment:
You can either:

- Wrap `err` in `OpenFileError`
- Return `fmt.Errorf("...", err)`